### PR TITLE
Add more heuristics for hiding obvious param hints

### LINKED ISSUE #3900: [Feature request] Hide obvious parameter name inlay hints
Sometimes, parameter names are simply called `value`. I don't think it's necessary to show those hints since they don't add anything. 

For example

image

Seeing the `value:` on each field really doesn't tell me anything that wasn't already obvious.

I'm wondering if it could be possible to have a list of words that could be removed from parameter names inlay hints. It could be a predefined list of words, but I wouldn't be surprised if a lot of people didn't like that. So having a list of user defined words is probably better.

I'd be interested in making a PR if you want. I might need some guidance to find where this is handled.

### PR DESCRIPTION
**Pull request by IceSentry**
This will now hide `value`, `pat`, `rhs` and `other`. These words were selected from the std because they are used in commonly used functions with only a single param and are obvious by their use.

It will also hide the hint if the passed param **starts** or end with the param_name. Maybe we could also split on '_' and check if one of the string is the param_name.

I think it would be good to also hide `bytes` if the type is `[u8; n]` but I'm not sure how to get the param type signature.

Closes #3900

### CODE REVIEW

**Comment by IceSentry** [comment]
I added a way to ignore `&mut ` and `&` in the argument_name, but I'm not sure if there is a more efficient way to check this.

File: crates/ra_ide/src/inlay_hints.rs [hunk 1]
```diff
@@ -235,8 +235,19 @@ fn should_show_param_hint(
     param_name: &str,
     argument: &ast::Expr,
 ) -> bool {
-    let argument_string = argument.syntax().to_string();
-    if param_name.is_empty() || argument_string.ends_with(param_name) {
+    let argument_string = {
```
Comment (matklad) [inline crates/ra_ide/src/inlay_hints.rs:238]: Indeed, this is better expressed as `if let ast::RefExpr(argument) = argument`.

Probably better expressed as a helper function:

```
fn remove_reference(exr: ast::Expr) -> ast::Expr
```

File: crates/ra_ide/src/inlay_hints.rs [hunk 2]
```diff
@@ -245,8 +256,14 @@ fn should_show_param_hint(
...
     };
+
     // avoid displaying hints for common functions like map, filter, etc.
-    if parameters_len == 1 && (param_name.len() == 1 || param_name == "predicate") {
+    // or other obvious words used in std
+    let is_obvious_param_name = match param_name {
+        "predicate" | "value" | "pat" | "rhs" | "other" => true,
+        _ => false,
+    };
+    if parameters_len == 1 && (param_name.len() == 1 || is_obvious_param_name) {
```
Comment (matklad) [inline crates/ra_ide/src/inlay_hints.rs:266]: Let's extract the `(param_name.len() == 1 || is_obvious_param_name)` into a 

```
fn is_obvious_param(param: &str) -> bool
```

function?

File: crates/ra_ide/src/inlay_hints.rs [hunk 3]
```diff
@@ -1,4 +1,4 @@
-//! FIXME: write short doc here
+//! This module defines multiple types of inlay hints and their visibility
```
Comment (matklad) [inline crates/ra_ide/src/inlay_hints.rs:1]: :heart_eyes:

File: crates/ra_ide/src/inlay_hints.rs (hunk 1 above)
Comment (IceSentry) [reply crates/ra_ide/src/inlay_hints.rs]: So, I tried to make a function that strips the Ref part, but I ended up with a token which I didn't know how to convert back to an ast::Expr. Instead I made a `fn is_argument_similar_to_param(argument, param_name) -> bool`. It makes the intent pretty explicit, but it's less reusable.

File: crates/ra_ide/src/inlay_hints.rs [hunk 4]
```diff
@@ -245,14 +244,32 @@ fn should_show_param_hint(
...
+    if parameters_len == 1 && is_obvious_param(param_name) {
         return false;
     }
-
     true
 }
 
+fn is_argument_similar_to_param(argument: &ast::Expr, param_name: &str) -> bool {
+    let argument_string = if let ast::Expr::RefExpr(ref_expr) = argument {
+        ref_expr.syntax().last_token().expect("RefExpr should have a last_token").to_string()
```
Comment (matklad) [inline crates/ra_ide/src/inlay_hints.rs:258]: This expect can in theory fire. I suggest doing this:

```
fn remove_ref(expr: ast::Expr) -> ast::Expr {
    if let ast::Expr::RefExpr(ref_expr) = &expr {
        if let Some(inner) = ref_expr.expr() {
            return expr;
        }
    }
    expr
}
```

Comment (IceSentry) [reply crates/ra_ide/src/inlay_hints.rs:258]: I didn't realize there was a `.expr()` on the RefExpr. I'll fix that!

**Comment by matklad** [comment]
bors r+

Thanks!

**Comment by lnicola** [comment]
I _think_ the test failure is spurious. Can someone retry this?

**Comment by edwin0cheng** [comment]
bors retry

**Pull request merged**